### PR TITLE
Implement Friends Demographics

### PR DESCRIPTION
**Pull request by tkgauri**
Implement [Get friend demographics](https://developers.line.biz/en/reference/messaging-api/#get-demographic).

This closes https://github.com/line/line-bot-sdk-java/issues/349.

### CODE REVIEW

File: line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingClient.java [hunk 1]
```diff
@@ -337,6 +338,13 @@
      */
     CompletableFuture<GetNumberOfFollowersResponse> getNumberOfFollowersResponse(String date);
 
+    /**
+     * Retrieves the demographic attributes for a bot's friends.
+     *
+     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-demographic">Get friends demographics</a>
+     */
+    CompletableFuture<GetFriendsDemographicsResponse> getFriendsDemographicsResponse();
```
Comment (kazuki-ma) [inline line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingClient.java]:
```suggestion
    CompletableFuture<GetFriendsDemographicsResponse> getFriendsDemographics();
```

`Response` is not needed for method signature.

File: line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingClient.java [hunk 2]
```diff
@@ -337,6 +338,13 @@
      */
     CompletableFuture<GetNumberOfFollowersResponse> getNumberOfFollowersResponse(String date);
 
+    /**
+     * Retrieves the demographic attributes for a bot's friends.
```
Comment (kazuki-ma) [inline line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingClient.java]:
```suggestion
     * Gets the demographic attributes for a bot's friends.
```

File: line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingClientImpl.java [hunk 3]
```diff
@@ -244,6 +245,11 @@
         return toFuture(retrofitImpl.issueLinkToken(userId));
     }
 
+    @Override
+    public CompletableFuture<GetFriendsDemographicsResponse> getFriendsDemographicsResponse() {
```
Comment (kazuki-ma) [inline line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingClientImpl.java]:
```suggestion
    public CompletableFuture<GetFriendsDemographicsResponse> getFriendsDemographics() {
```

File: line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingService.java [hunk 4]
```diff
@@ -329,6 +330,12 @@
     @POST("v2/bot/user/{userId}/linkToken")
     Call<IssueLinkTokenResponse> issueLinkToken(@Path("userId") String userId);
 
+    /**
+     * Retrieves the demographic attributes for a bot's friends.
```
Comment (kazuki-ma) [inline line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingService.java]:
```suggestion
     * Gets the demographic attributes for a bot's friends.
```

File: line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+@Builder
+@JsonDeserialize(builder = GetFriendsDemographicsResponseBuilder.class)
+public class GetFriendsDemographicsResponse {
+
+    @Value
+    @Builder
+    @JsonDeserialize(builder = GenderWisePercentage.GenderWisePercentageBuilder.class)
+    static class GenderWisePercentage {
+        String gender;
+        Double percentage;
```
Comment (kazuki-ma) [inline line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java]:
```suggestion
        double percentage;
```

if it is never null.

File: line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+@Value
+@Builder
+@JsonDeserialize(builder = GetFriendsDemographicsResponseBuilder.class)
+public class GetFriendsDemographicsResponse {
+
+    @Value
+    @Builder
+    @JsonDeserialize(builder = GenderWisePercentage.GenderWisePercentageBuilder.class)
+    static class GenderWisePercentage {
+        String gender;
```
Comment (kazuki-ma) [inline line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java]: Please add comment. To user can know we choose `M` or `F`, `male` or `female` etc..

File: line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java (hunk 5 above)
Comment (kazuki-ma) [reply line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java]: And please add comment to `[0.0-100.0]`.

File: line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        @JsonPOJOBuilder(withPrefix = "")
+        public static class  SubscriptionWisePercentageBuilder{
+            // Filled by lombok
+        }
+    }
+
+    /**
+     * true if friend demographic information is available.
+     */
+    Boolean available;
```
Comment (kazuki-ma) [inline line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java]:
```suggestion
    bool available;
```

File: line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        public static class GenderWisePercentageBuilder{
+            // Filled by lombok
+        }
+    }
+
+    @Value
+    @Builder
+    @JsonDeserialize(builder = AgeWisePercentage.AgeWisePercentageBuilder.class)
+    static class AgeWisePercentage {
+        String age;
```
Comment (kazuki-ma) [inline line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java]: Could you add example value of it?

File: line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        public static class  AreaWisePercentageBuilder{
+            // Filled by lombok
+        }
+    }
+
+    @Value
+    @Builder
+    @JsonDeserialize(builder = OSWisePercentage.OSWisePercentageBuilder.class)
+    static class OSWisePercentage {
+        String appType;
```
Comment (kazuki-ma) [inline line-bot-model/src/main/java/com/linecorp/bot/model/response/GetFriendsDemographicsResponse.java]: Example value please.

File: line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingService.java (hunk 4 above)
Comment (kazuki-ma) [reply line-bot-api-client/src/main/java/com/linecorp/bot/client/LineMessagingService.java]: Sorry, official document write so. It's ok.

image

**Comment by kazuki-ma** [comment]
I'm taking over this task from @tkgauri and blush up somecode.

@imasahiro, @koya3 Could you review it?

Thank you.

File: line-bot-model/src/main/java/com/linecorp/bot/model/response/demographics/AgeRange.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    @JsonValue
+    String getJsonRawValue();
+
+    /**
+     * Returns true if the given objects are considered equivalent.
+     *
+     * <p>That means from&amp;to are equal. That's includes case of both are UNKNOWN.
+     */
+    boolean equals(Object another);
```
Comment (imasahiro) [inline line-bot-model/src/main/java/com/linecorp/bot/model/response/demographics/AgeRange.java]: I think, we don't need to add this method to this interface.

File: line-bot-model/src/main/java/com/linecorp/bot/model/response/demographics/GetFriendsDemographicsResponse.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import lombok.Builder.Default;
+import lombok.Value;
+
+@Value
+@Builder
+@JsonDeserialize(builder = GetFriendsDemographicsResponse.GetFriendsDemographicsResponseBuilder.class)
+public class GetFriendsDemographicsResponse {
+
+    /**
+     * true if friend demographic information is available.
```
Comment (imasahiro) [inline line-bot-model/src/main/java/com/linecorp/bot/model/response/demographics/GetFriendsDemographicsResponse.java]: nit.
`{@literal true} if friend demographic information is available.`

**Comment by kazuki-ma** [comment]
Updated.

**Comment by kazuki-ma** [comment]
Updated. Please review again.

**Pull request merged**